src/Interpreters/MergeJoin.cpp

```diff
@@ -433,6 +434,27 @@ MergeJoin::MergeJoin(std::shared_ptr<TableJoin> table_join_, const Block & right
     , max_rows_in_right_block(table_join->maxRowsInRightBlock())
     , max_files_to_merge(table_join->maxFilesToMerge())
 {
+    const auto & key_names_left = table_join->keyNamesLeft();
+    const auto & key_names_right = table_join->keyNamesRight();
+
+    bool can_cast_key_columns = JoinCommon::canCastJoinColumns(left_sample_block, *table_join);
+    if (can_cast_key_columns)
+    {
+        auto cols_to_convert = JoinCommon::getJoinColumnsNeedCast(
+            left_sample_block, table_join->keyNamesLeft(), right_sample_block, key_names_right);
+
+        /// Casted columns will be stored in separate columns with unique generated names
+        String uuid_string = UUIDHelpers::generateV4().toUnderType().toHexString();
```

Comment:
UUIDs in column names can be confusing. Maybe we can use names like `cast(UInt8, left_table.key_column)`? If you're using the `cast` function, you should get these names in a natural way.